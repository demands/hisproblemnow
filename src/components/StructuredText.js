import React, {PropTypes} from 'react';

const StructuredText = (props) => (
  <div className='structured-text'>
    {props.value.map((block, i) => {
      if (block.type === 'paragraph') {
        return (
          <p className='block' key={i}>
            {block.text.split('\n').map((segment, i) => 
              <span key={i}>
                {segment}<br />
              </span>
            )}
          </p>
        )
      }
    })}
  </div>
);

export default StructuredText;
