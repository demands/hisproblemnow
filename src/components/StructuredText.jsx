import React, { PropTypes } from 'react';
import _ from 'lodash';

function StructuredText({ value }) {
  const blocks = [];
  let listItems = [];

  /* group contiguous list items */
  (value || []).forEach((block) => {
    if (block.type === 'list-item') {
      listItems.push(block);
    } else {
      if (listItems.length > 0) {
        blocks.push({
          type: 'unordered-list',
          listItems,
        });
        listItems = [];
      }
      blocks.push(block);
    }
  });
  if (listItems.length > 0) {
    blocks.push({
      type: 'unordered-list',
      listItems,
    });
    listItems = [];
  }

  return (
    <div className="structured-text">
      {blocks.map((block, blockKey) => {
        if (block.type === 'unordered-list') {
          return (<ul key={blockKey}>
            {block.listItems.map((listItem, listItemKey) =>
              <Block key={listItemKey} {...listItem} />,
            )}
          </ul>);
        }
        return <Block key={blockKey} {...block} />;
      })}
    </div>
  );
}

StructuredText.propTypes = {
  value: PropTypes.array,
};

function Block({ type, spans, text }) {
  const segments = [];
  let index = 0;
  let tag;

  spans.forEach((span) => {
    if (span.start !== index) {
      segments.push({
        text: text.slice(index, span.start),
        type: 'span',
      });
    }
    segments.push({
      text: text.slice(span.start, span.end),
      type: span.type,
      data: span.data,
    });
    index = span.end;
  });
  if (index < text.length - 1) {
    segments.push({
      text: text.slice(index),
      type: 'span',
    });
  }

  switch (type) {
    case 'list-item':
      tag = 'li';
      break;
    case 'paragraph':
    default:
      tag = 'p';
      break;
  }
  return React.createElement(tag, {},
    segments.map((segment, i) => {
      switch (segment.type) {
        case 'span':
          return (<span key={i}><Segment text={segment.text} /></span>);
        case 'strong':
          return (<b key={i}><Segment text={segment.text} /></b>);
        case 'hyperlink':
          return (<a
            key={i}
            href={_.get(segment, 'data.value.url')}
            target="_blank"
            rel="noopener noreferrer"
          >
            <Segment text={segment.text} />
          </a>);
        default:
          return (<tt key={i}>{JSON.stringify(segment)}</tt>);
      }
    }),
  );
}

Block.propTypes = {
  type: PropTypes.string.isRequired,
  spans: PropTypes.array.isRequired,
  text: PropTypes.string.isRequired,
};

function Segment({ text }) {
  const lines = text.split('\n');
  return (
    <span>
      {lines.map((line, i) =>
        <span key={i}>
          {line}{i < lines.length - 1 && <br />}
        </span>,
      )}
    </span>
  );
}

Segment.propTypes = {
  text: PropTypes.string.isRequired,
};

export default StructuredText;
