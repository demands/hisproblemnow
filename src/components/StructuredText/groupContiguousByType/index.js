/* @flow */

interface Element {
  type: string,
}

export default function groupContiguousByType(elements: Element[]): Element[][] {
  const groups = [];
  let groupStart = 0;
  let element = elements[0];
  let lastType = element && element.type;

  for (let i = 0; i <= elements.length; i += 1) {
    element = elements[i];
    if (!element || element.type !== lastType) {
      groups.push(elements.slice(groupStart, i));
      groupStart = i;
      lastType = element && element.type;
    }
  }

  return groups;
}
