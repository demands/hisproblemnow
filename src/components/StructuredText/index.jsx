import React, { PropTypes } from 'react';
import _ from 'lodash';

import groupContiguousByType from './groupContiguousByType';

const TAG_BY_BLOCK_TYPE = {
  'list-item': 'li',
  'o-list-item': 'li',
  paragraph: 'p',
};

const WRAPPER_TAG_BY_BLOCK_TYPE = {
  'list-item': 'ul',
  'o-list-item': 'ol',
};

function StructuredText({ value }) {
  const groups = groupContiguousByType(value || []);

  return (
    <div className="structured-text">
      {_.flatten(groups.map((group, groupKey) => {
        if (group.length === 0) return [];

        const groupType = group[0].type;
        const groupBlocks = group.map((block, blockKey) => <Block
          key={`group-${groupKey}|block-${blockKey}`}
          {...block}
        />);

        if (WRAPPER_TAG_BY_BLOCK_TYPE[groupType]) {
          const tag = WRAPPER_TAG_BY_BLOCK_TYPE[groupType];
          return React.createElement(tag, { key: `group-${groupKey}` }, groupBlocks);
        }

        return groupBlocks;
      }))}
    </div>
  );
}

StructuredText.propTypes = {
  value: PropTypes.array,
};

function Block({ type, spans, text }) {
  const segments = [];
  const tag = TAG_BY_BLOCK_TYPE[type] || 'p';
  let index = 0;

  spans.forEach((span) => {
    if (span.start !== index) {
      segments.push({
        text: text.slice(index, span.start),
        type: 'span',
      });
    }
    segments.push({
      text: text.slice(span.start, span.end),
      type: span.type,
      data: span.data,
    });
    index = span.end;
  });

  if (index < text.length) {
    segments.push({
      text: text.slice(index),
      type: 'span',
    });
  }

  return React.createElement(tag, {},
    segments.map((segment, i) => {
      switch (segment.type) {
        case 'span':
          return (<span key={i}><Segment text={segment.text} /></span>);
        case 'strong':
          return (<b key={i}><Segment text={segment.text} /></b>);
        case 'hyperlink':
          return (<a
            key={i}
            href={_.get(segment, 'data.value.url')}
            target="_blank"
            rel="noopener noreferrer"
          >
            <Segment text={segment.text} />
          </a>);
        default:
          return (<tt key={i}>{JSON.stringify(segment)}</tt>);
      }
    }),
  );
}

Block.propTypes = {
  type: PropTypes.string.isRequired,
  spans: PropTypes.array.isRequired,
  text: PropTypes.string.isRequired,
};

function Segment({ text }) {
  const lines = text.split('\n');
  return (
    <span>
      {lines.map((line, i) =>
        <span key={i}>
          {line}{i < lines.length - 1 && <br />}
        </span>,
      )}
    </span>
  );
}

Segment.propTypes = {
  text: PropTypes.string.isRequired,
};

export default StructuredText;
