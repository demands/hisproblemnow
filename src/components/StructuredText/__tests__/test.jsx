/* eslint-env jest */

import React from 'react';
import renderer from 'react-test-renderer';
import StructuredText from '../index';

it('does not render when provided with no input', () => {
  const component = renderer.create(<StructuredText />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders a paragraph', () => {
  const component = renderer.create(
    <StructuredText
      value={[
        {
          type: 'paragraph',
          text: 'Lorem ipsum dolor sit amet',
          spans: [],
        },
      ]}
    />,
  );
  expect(component.toJSON()).toMatchSnapshot();
});

it('renders multiple paragraphs', () => {
  const component = renderer.create(
    <StructuredText
      value={[
        {
          type: 'paragraph',
          text: 'Lorem ipsum dolor sit amet',
          spans: [],
        },
        {
          type: 'paragraph',
          text: 'Consectitur adipiscing elit',
          spans: [],
        },
      ]}
    />,
  );
  expect(component.toJSON()).toMatchSnapshot();
});

it('renders bold text', () => {
  const component = renderer.create(
    <StructuredText
      value={[
        {
          type: 'paragraph',
          text: 'Lorem ipsum dolor sit amet!',
          spans: [{ start: 12, end: 26, type: 'strong' }],
        },
      ]}
    />,
  );
  expect(component.toJSON()).toMatchSnapshot();
});

it('renders hyperlinks', () => {
  const component = renderer.create(
    <StructuredText
      value={[
        {
          type: 'paragraph',
          text: 'Lorem ipsum dolor sit amet!',
          spans: [
            {
              start: 0,
              end: 5,
              type: 'hyperlink',
              data: {
                value: {
                  url: 'http://example.com',
                },
              },
            },
          ],
        },
      ]}
    />,
  );
  expect(component.toJSON()).toMatchSnapshot();
});

it('renders unordered lists', () => {
  const component = renderer.create(
    <StructuredText
      value={[
        {
          type: 'list-item',
          text: 'Lorem ipsum dolor sit amet!',
          spans: [],
        },
        {
          type: 'list-item',
          text: 'Sed do eiusmod tempor incidunt.',
          spans: [],
        },
        {
          type: 'list-item',
          text: 'Ut labore et dolore.',
          spans: [],
        },
      ]}
    />,
  );
  expect(component.toJSON()).toMatchSnapshot();
});

it('renders ordered lists', () => {
  const component = renderer.create(
    <StructuredText
      value={[
        {
          type: 'o-list-item',
          text: 'Lorem ipsum dolor sit amet!',
          spans: [],
        },
        {
          type: 'o-list-item',
          text: 'Sed do eiusmod tempor incidunt.',
          spans: [],
        },
        {
          type: 'o-list-item',
          text: 'Ut labore et dolore.',
          spans: [],
        },
      ]}
    />,
  );
  expect(component.toJSON()).toMatchSnapshot();
});

it('renders combinations of structured text elements', () => {
  const component = renderer.create(
    <StructuredText
      value={[
        {
          type: 'o-list-item',
          text: 'Lorem ipsum dolor sit amet!',
          spans: [],
        },
        {
          type: 'list-item',
          text: 'Sed do eiusmod tempor incidunt.',
          spans: [],
        },
        {
          type: 'paragraph',
          text: 'Ut labore et dolore.',
          spans: [],
        },
      ]}
    />,
  );
  expect(component.toJSON()).toMatchSnapshot();
});
