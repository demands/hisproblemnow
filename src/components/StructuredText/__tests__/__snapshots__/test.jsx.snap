exports[`test does not render when provided with no input 1`] = `
<div
  className="structured-text" />
`;

exports[`test renders a paragraph 1`] = `
<div
  className="structured-text">
  <p>
    <span>
      <span>
        <span>
          Lorem ipsum dolor sit amet
        </span>
      </span>
    </span>
  </p>
</div>
`;

exports[`test renders bold text 1`] = `
<div
  className="structured-text">
  <p>
    <span>
      <span>
        <span>
          Lorem ipsum 
        </span>
      </span>
    </span>
    <b>
      <span>
        <span>
          dolor sit amet
        </span>
      </span>
    </b>
    <span>
      <span>
        <span>
          !
        </span>
      </span>
    </span>
  </p>
</div>
`;

exports[`test renders combinations of structured text elements 1`] = `
<div
  className="structured-text">
  <ol>
    <li>
      <span>
        <span>
          <span>
            Lorem ipsum dolor sit amet!
          </span>
        </span>
      </span>
    </li>
  </ol>
  <ul>
    <li>
      <span>
        <span>
          <span>
            Sed do eiusmod tempor incidunt.
          </span>
        </span>
      </span>
    </li>
  </ul>
  <p>
    <span>
      <span>
        <span>
          Ut labore et dolore.
        </span>
      </span>
    </span>
  </p>
</div>
`;

exports[`test renders hyperlinks 1`] = `
<div
  className="structured-text">
  <p>
    <a
      href="http://example.com"
      rel="noopener noreferrer"
      target="_blank">
      <span>
        <span>
          Lorem
        </span>
      </span>
    </a>
    <span>
      <span>
        <span>
           ipsum dolor sit amet!
        </span>
      </span>
    </span>
  </p>
</div>
`;

exports[`test renders multiple paragraphs 1`] = `
<div
  className="structured-text">
  <p>
    <span>
      <span>
        <span>
          Lorem ipsum dolor sit amet
        </span>
      </span>
    </span>
  </p>
  <p>
    <span>
      <span>
        <span>
          Consectitur adipiscing elit
        </span>
      </span>
    </span>
  </p>
</div>
`;

exports[`test renders ordered lists 1`] = `
<div
  className="structured-text">
  <ol>
    <li>
      <span>
        <span>
          <span>
            Lorem ipsum dolor sit amet!
          </span>
        </span>
      </span>
    </li>
    <li>
      <span>
        <span>
          <span>
            Sed do eiusmod tempor incidunt.
          </span>
        </span>
      </span>
    </li>
    <li>
      <span>
        <span>
          <span>
            Ut labore et dolore.
          </span>
        </span>
      </span>
    </li>
  </ol>
</div>
`;

exports[`test renders unordered lists 1`] = `
<div
  className="structured-text">
  <ul>
    <li>
      <span>
        <span>
          <span>
            Lorem ipsum dolor sit amet!
          </span>
        </span>
      </span>
    </li>
    <li>
      <span>
        <span>
          <span>
            Sed do eiusmod tempor incidunt.
          </span>
        </span>
      </span>
    </li>
    <li>
      <span>
        <span>
          <span>
            Ut labore et dolore.
          </span>
        </span>
      </span>
    </li>
  </ul>
</div>
`;
